nemo/collections/multimodal/data/energon/base.py

```diff
@@ -0,0 +1,377 @@
+        self.data_sampler.prev_consumed_samples = consumed_samples
+        logging.info(f"Multimodal dataloader load state dict with consumed_samples {consumed_samples}")
+        update_num_microbatches(
+            consumed_samples=consumed_samples,
+            consistency_check=False,
+        )
+
+
+class SequentialMegatronSampler(MegatronDataSampler):
+    """
+    A data sampler for sequential sampling in Megatron, designed to handle large datasets efficiently.
+
+    This class extends the MegatronDataSampler to support sequential sampling for large datasets.
+    It includes functionality for handling micro-batches and tracking consumed samples across training steps.
+
+    Attributes:
+    seq_len (int): The sequence length for each sample.
+    micro_batch_size (int): The number of samples in each micro-batch.
+    init_consumed_samples (int): The initial number of samples that have been consumed (used for resuming training).
+    prev_consumed_samples (int): Tracks the number of consumed samples before the current step.
+    if_first_step (int): Flag to indicate if it's the first training step.
+    prev_global_batch_size (Optional[int]): The global batch size from the previous step.
+    init_global_step (int): The initial global step at the start of training.
+    """
+
+    def __init__(self, seq_len: int, micro_batch_size: int = 4, init_consumed_samples: int = 0, init_global_step=0):
+        """
+        Initialize the SequentialMegatronSampler.
+
+        Parameters:
+        seq_len (int): The sequence length for each sample.
+        micro_batch_size (int, optional): The number of samples in each micro-batch. Defaults to 4.
+        init_consumed_samples (int, optional): The initial number of samples that have been consumed. Defaults to 0.
+        init_global_step (int, optional): The initial global step at the start of training. Defaults to 0.
+        """
+        super().__init__(seq_len=seq_len, micro_batch_size=micro_batch_size)
+        self.seq_len = seq_len
+        self.micro_batch_size = micro_batch_size
+        self.init_consumed_samples = init_consumed_samples
+        self.prev_consumed_samples = self.init_consumed_samples
+        self.if_first_step = 0
+        self.prev_global_batch_size = None
+        self.init_global_step = init_global_step
+
+    def setup(self, global_rank: int) -> None:
+        """
+        Setup the sampler for the given global rank.
+
+        This method initializes the micro-batch calculator based on the global rank and batch size.
+        It ensures that the data is properly distributed across multiple devices in a distributed setting.
+
+        Parameters:
+        global_rank (int): The global rank of the current process in the distributed training setup.
+        """
+        from nemo.lightning.data import setup_microbatch_calculator
+
+        self.global_batch_size = self.micro_batch_size * parallel_state.get_data_parallel_world_size()
```

Comment:
allow user to set global_batch_size? some might want to use something larger for grad accumulation

Reply:
Thanks for point out. Allowed the user to specify custom global batch sizes for grad accumulation.